File: No_1579/Remove Max Number of Edges to Keep Graph Fully Traversable.cpp
class UnionFind {
    vector<int> groups, heights;
    int components;
public:
    UnionFind(int n): groups(n+1), heights(n+1, 1), components(n){}
    
    int Union(int i, int j) {
        int root1 = Find(i);
        int root2 = Find(j);
       
        if (root1 == root2) return 0;
        
        if (heights[root1] < heights[root2]) {
            groups[root1] = root2;
            heights[root2] = max(heights[root2], heights[root1] + 1);
        } else {
            groups[root2] = root1;    
            heights[root1] = max(heights[root1], heights[root2] + 1);
        }
        
        components--;
        return 1;
    }
    
    int Find(int i) {
        if (groups[i] == 0)
            return i;
        
        return groups[i] = Find(groups[i]);
    }
    
    bool isConnected() {
        return components == 1;
    }
};

class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        UnionFind Alice(n), Bob(n);
        
        int requiredEdges = 0;
        for (auto e : edges) {
            if (e[0] == 3)
                if (Alice.Union(e[1], e[2]) and Bob.Union(e[1], e[2]))
                    requiredEdges++;
        }
        
        for (auto e: edges) {
            if (e[0] == 1)
                requiredEdges += Alice.Union(e[1], e[2]);
            else if (e[0] == 2)
                requiredEdges += Bob.Union(e[1], e[2]);
            
        }
        
        if (Alice.isConnected() && Bob.isConnected())
            return edges.size() - requiredEdges;
        
        return -1;
    }
};

File: No_1579/Remove Max Number of Edges to Keep Graph Fully Traversable.py
class UnionFind():
    def __init__(self, n):
        self.__groups = [0] * (n + 1)
        self.__heights = [1] * (n + 1)
        self.components = n
    
    def Union(self, i, j):
        root1 = self.Find(i)
        root2 = self.Find(j)
    
        if root1 == root2:
            return 0
        
        if self.__heights[root1] < self.__heights[root2]:
            self.__groups[root1] = root2
            self.__heights[root2] = max(self.__heights[root2], self.__heights[root1] + 1)
        else:
            self.__groups[root2] = root1
            self.__heights[root1] = max(self.__heights[root1], self.__heights[root2] + 1)
    
        self.components -= 1
        return 1
    
    def Find(self, i):
        if self.__groups[i] == 0:
            return i
        
        self.__groups[i] = self.Find(self.__groups[i])
        return self.__groups[i]
    
    def is_connected(self):
        return self.components == 1
        

class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        Alice = UnionFind(n)
        Bob = UnionFind(n)
        
        requiredEdges = 0
        for e in edges:
            if e[0] == 3:
                if Alice.Union(e[1], e[2]) and Bob.Union(e[1], e[2]):
                    requiredEdges += 1
                    
        for e in edges:
            if e[0] == 1:
                requiredEdges += Alice.Union(e[1], e[2])
            elif e[0] == 2:
                requiredEdges += Bob.Union(e[1], e[2])
                
        if Alice.is_connected() and Bob.is_connected():
            return len(edges) - requiredEdges
        
        return -1
        